### PR TITLE
Add `.webappignore` and `.funcignore` Support for Service Packaging

### PR DESCRIPTION
**Pull request by jongio**
Fixes: https://github.com/Azure/azure-dev/issues/1039 https://github.com/Azure/azure-dev/issues/4215

### Description:

This PR introduces support for `.webappignore` and `.funcignore` files to the service packaging process within the Azure Developer CLI (azd). These files allow users to specify files and directories that should be excluded from the deployment package for App Service and Function Apps, providing fine-grained control at the service level.

#### Key Updates:

1. **`.webappignore` and `.funcignore` File Support:**
   - Implemented the ability to define `.webappignore` for App Services and `.funcignore` for Azure Functions within individual service directories.
   - If a `.webappignore` or `.funcignore` file exists in a service directory, the packaging process will follow its rules, allowing granular control over which files are included or excluded in the service's zip archive.
   - These ignore files ensure that users can customize file exclusions while packaging for App Services and Azure Functions independently.

2. **Updated Exclusion Logic:**
   - By default, the packaging process excludes directories like `__pycache__`, `.venv`, and `node_modules` unless a `.webappignore` or `.funcignore` file is present, which can override these default exclusions.
   - The packaging process respects these ignore files:
     - If a `.webappignore` file is present in an App Service, directories such as `node_modules` or `__pycache__` can be included or excluded as specified.
     - For Azure Functions, `.funcignore` can similarly be used to specify inclusion/exclusion rules.
   - If no `.webappignore` or `.funcignore` file is present, default exclusions apply for Python (`__pycache__`, `.venv`) and Node.js (`node_modules`).

3. **Test Coverage:**
   - Extensive tests have been added to validate the behavior under different scenarios:
     - Packaging with and without `.webappignore` and `.funcignore` files.
     - Ensuring correct exclusions of `__pycache__`, `.venv`, and `node_modules` based on the presence or absence of the ignore files.
     - Ensuring service-specific `.webappignore` and `.funcignore` behavior is correctly respected.
     - Verifying that specific files such as `logs/log.txt` can be excluded based on the ignore file configuration.

4. **Sample Project Structure:**
   - Updated the sample project structure used in tests to include services representing both App Services and Azure Functions (Python and Node.js examples).
   - Added scenarios where `.webappignore` files explicitly include or exclude `node_modules` or `__pycache__` based on real-world usage patterns.

5. **Documentation:**
   - **TODO:** Update documentation to include guidance on how to use `.webappignore` and `.funcignore` files with examples for both Python and Node.js projects.

### Testing:

All test cases have been successfully executed, covering a broad range of scenarios to ensure expected packaging behavior. The tests ensure that both `.webappignore` and `.funcignore` files are respected and that default exclusions are correctly applied when no ignore file is present.

### Additional Notes:

- The change is backward-compatible and only affects users who introduce `.webappignore` or `.funcignore` files into their services.
- This feature gives users more control over the contents of their deployment packages, allowing for more streamlined and efficient deployments.

### How to Review:

- Review the changes to the service packaging logic, focusing on how `.webappignore` and `.funcignore` files influence the exclusion logic.
- Ensure that test cases provide full coverage for both App Service and Azure Function scenarios.
- **TODO:** Ensure documentation updates reflect the new functionality for users.

### CODE REVIEW

File: cli/azd/pkg/project/framework_service_npm.go [hunk 1]
```diff
@@ -130,11 +130,7 @@ func (np *npmProject) Package(
 	if err := buildForZip(
 		packageSource,
 		packageDest,
-		buildForZipOptions{
-			excludeConditions: []excludeDirEntryCondition{
-				excludeNodeModules,
-			},
-		}); err != nil {
+		buildForZipOptions{}); err != nil {
```
Comment (jongio) [inline cli/azd/pkg/project/framework_service_npm.go:133]: These exclusions where moved to project_utils, because they are now dependent on the existance of .zipignore

File: cli/azd/pkg/project/framework_service_python.go [hunk 2]
```diff
@@ -130,12 +130,7 @@ func (pp *pythonProject) Package(
 	if err := buildForZip(
 		packageSource,
 		packageDest,
-		buildForZipOptions{
-			excludeConditions: []excludeDirEntryCondition{
-				excludeVirtualEnv,
-				excludePyCache,
-			},
-		}); err != nil {
+		buildForZipOptions{}); err != nil {
```
Comment (jongio) [inline cli/azd/pkg/project/framework_service_python.go:133]: These exclusions where moved to project_utils, because they are now dependent on the existance of .zipignore

**Comment by pamelafox** [comment]
This seems helpful, per discussions on the other PR! My main design question would be why a separate file versus an --ignore in the azure.yaml. Is it easier to define globs that way? Does YAML have annoying syntax restrictions?

**Comment by vhvb1989** [comment]
> This seems helpful, per discussions on the other PR! My main design question would be why a separate file versus an --ignore in the azure.yaml. Is it easier to define globs that way? Does YAML have annoying syntax restrictions?

One of the reasons was for making this new ignore-file recognizable beyond azd; so you could use it when using web-app-deploy action (https://github.com/Azure/webapps-deploy) or your own tools.  @jongio was going to check with App Service if there was something already for it and if not, it might be worth it to have it beyond azd; similar to .dockerignore, .npmignore, etc.  Then azd would just honor this file like any other tool would.

**Comment by jongio** [comment]
> > This seems helpful, per discussions on the other PR! My main design question would be why a separate file versus an --ignore in the azure.yaml. Is it easier to define globs that way? Does YAML have annoying syntax restrictions?
> 
> One of the reasons was for making this new ignore-file recognizable beyond azd; so you could use it when using web-app-deploy action (https://github.com/Azure/webapps-deploy) or your own tools. @jongio was going to check with App Service if there was something already for it and if not, it might be worth it to have it beyond azd; similar to .dockerignore, .npmignore, etc. Then azd would just honor this file like any other tool would.

Yes, I'm in talks with them on this.

**Comment by pamelafox** [comment]
Ah, cool, that design makes sense then!

File: cli/azd/pkg/project/framework_service_python.go (hunk 2 above)
Comment (vhvb1989) [inline cli/azd/pkg/project/framework_service_python.go:133]: each `framework_service_language` should continue to provide a default `excludeConditions` for when no .zipignore is defined in the project.

Maybe you can call a method here from the `dotignore` package and pass the default excluding files.  The method will check if .zipignore exists. If no, it uses the default rules passed by the `framework_service_language`. If yes, it uses the rules from the .zipignore

File: cli/azd/pkg/project/project_utils.go [hunk 3]
```diff
@@ -48,12 +57,46 @@ type buildForZipOptions struct {
...
+	if err != nil && !os.IsNotExist(err) {
+		return fmt.Errorf("reading .zipignore files: %w", err)
+	}
+
+	// Determine if the .zipignore file exists at the root of the service
+	zipIgnoreExists := len(ignoreMatchers) > 0
+
+	// Conditionally exclude virtual environments, __pycache__, and node_modules only if .zipignore doesn't exist
+	if !zipIgnoreExists {
+		options.excludeConditions = append(options.excludeConditions, excludeVirtualEnv, excludePyCache, excludeNodeModules)
```
Comment (vhvb1989) [inline cli/azd/pkg/project/project_utils.go:75]: the list of things to exclude here should come from the `framework_service_language`.  This util method should not be the source of truth for that.

**Review by vhvb1989 — COMMENTED** [review]
few comments

**Comment by jongio** [comment]
Comment from @Petermarcu 

https://github.com/Azure/azure-dev/pull/4146#issuecomment-2319020209

Is zip too specific and this is more of a ignore on deploy in azd terms? deployignore? Will we use a similar concept if the output is not in the form of a zip file in the future? Even just a lose folder of files that is ready to deploy?


@ellismg - You mentioned first mentioned .zipdeployignore, we landed on .zipignore because it's not necessarily part of the deploy action, instead it is part of packaging.

We want the file name to be generic enough to be used by other app service related tooling.  In the same vein as .funcignore, .dockerignore, should we name this .webappignore?

**Comment by Petermarcu** [comment]
Is this only going to be respected for app service services?

**Comment by jongio** [comment]
> Is this only going to be respected for app service services?

Yes, most of the other services already have .*ignore files, .funcignore, .dockerignore

**Comment by jongio** [comment]
Related: https://github.com/Azure/azure-dev/issues/4215

**Comment by jongio** [comment]
> We want the file name to be generic enough to be used by other app service related tooling. In the same vein as .funcignore, .dockerignore, should we name this .webappignore?

@ellismg - I've confirmed with the web app and functions teams that they would prefer `.webappignore`.  If no objections, I think we should move forward with that.  Agree?

**Comment by ellismg** [comment]
> @ellismg - I've confirmed with the web app and functions teams that they would prefer .webappignore. If no objections, I think we should move forward with that. Agree?

Sounds great to me, @jongio!

File: cli/azd/pkg/project/service_target.go [hunk 4]
```diff
@@ -94,6 +94,19 @@ type ServiceTarget interface {
...
+func GetIgnoreFileNameByKind(kind ServiceTargetKind) string {
+	switch kind {
+	case AzureFunctionTarget:
+		return ".funcignore"
+	case AppServiceTarget:
+		return ".webappignore"
+	default:
+		return ""
+	}
+}
```
Comment (jongio) [inline cli/azd/pkg/project/service_target.go:108]: I debated on whether or not we should have a helper method like this, or add this to the interface, but since only a few of the service targets will have this capability I figured we should add it as a helper method to not mess with the interface of the service target. But I could be convinced the other direction, depending on how you want to design this.

**Comment by jongio** [comment]
@wbreza @vhvb1989 @ellismg - Could you have a look at the updated code and PR description with all the changes?  I'm in discussion with the appservice team and the approach with .webappignore and .funcignore are aligned with them.

@pamelafox - Once we get a build could you verify this works as expected in your projects with and without .webappignore files?  (You'll obviously need to create the .webappignore file in the project this impacts)

@v-xuto - Can you please run manual regression / smoke tests on all the azdev ToDo templates and ensure that everything passes?

Thanks!

**Comment by kristenwomack** [comment]
> @wbreza @vhvb1989 @ellismg - Could you have a look at the updated code and PR description with all the changes? I'm in discussion with the appservice team and the approach with .webappignore and .funcignore are aligned with them.
> 
> @pamelafox - Once we get a build could you verify this works as expected in your projects with and without .webappignore files? (You'll obviously need to create the .webappignore file in the project this impacts)
> 
> @v-xuto - Can you please run manual regression / smoke tests on all the azdev ToDo templates and ensure that everything passes?
> 
> Thanks!

Do you also want to run the azdev ToDo templates through the validation pipeline with this change? cc @anfibiacreativa

**Comment by v-xuto** [comment]
> @v-xuto - Can you please run manual regression / smoke tests on all the azdev ToDo templates and ensure that everything passes?
> 

Got it. Testing.

File: cli/azd/pkg/project/service_target.go (hunk 4 above)
Comment (weikanglim) [reply cli/azd/pkg/project/service_target.go:108]: I like flexibility -- simple functions are great for that purpose. I'm okay here.

File: cli/azd/pkg/project/project_utils.go [hunk 5]
```diff
@@ -48,17 +50,43 @@ type buildForZipOptions struct {
...
-// details for each language which should not be ever copied.
-func buildForZip(src, dst string, options buildForZipOptions) error {
+// buildForZip is used by projects to prepare a directory for
+// zipping, excluding files based on the ignore file and other conditions.
+func buildForZip(src, dst string, options buildForZipOptions, serviceConfig *ServiceConfig) error {
+	// Lookup the appropriate ignore file name based on the service kind (Host)
+	ignoreFileName := GetIgnoreFileNameByKind(serviceConfig.Host)
+
+	// Read and honor the specified ignore file if it exists
+	ignoreMatcher, err := dotignore.ReadDotIgnoreFile(src, ignoreFileName)
```
Comment (weikanglim) [inline cli/azd/pkg/project/project_utils.go:60]: Are we also thinking about supporting `*.ignore` files in any recursive directory of the service?

```
- funcapp/
   - .funcignore
   - src/
     - otherThing/
       - .funcignore # define ignore files relative to this path
```

I think we would honor both ignore files.

If we're doing this, I think we should *consider* implementing the correct recursion logic in `project` before generalizing it as a `dotignore` package -- I would like to see the design matured out before we push it up to an app library package in `azd`.

**Review by weikanglim — COMMENTED** [review]
Don't have strong feelings yet on the PR, have an overall question about whether we're thinking about recursive directories.

**Review by weikanglim — COMMENTED** [review]
> Are we also thinking about supporting `*.ignore` files in any recursive directory of the service?

> ```
- funcapp/
   - .funcignore
   - src/
     - otherThing/
       - .funcignore # define ignore files relative to this path
```

> I think we would honor both ignore files.

> If we're doing this, I think we should *consider* implementing the correct recursion logic in `project` before generalizing it as a `dotignore` package -- I would like to see the design matured out before we push it up to an app library package in `azd`.

File: cli/azd/pkg/project/project_utils.go (hunk 5 above)
Comment (weikanglim) [reply cli/azd/pkg/project/project_utils.go:60]: @ellismg had implemented `.dockerignore` recently -- wonder if he has thoughts here.

File: cli/azd/pkg/project/project_utils.go [hunk 6]
```diff
@@ -9,34 +9,36 @@ import (
...
 
+	// Close the zip file and return the path
 	if err := zipFile.Close(); err != nil {
-		// may fail but, again, we'll do our best to cleanup here.
 		os.Remove(zipFile.Name())
 		return "", err
 	}
 
-	return zipFile.Name(), nil
+	return filePath, nil
```
Comment (wbreza) [inline cli/azd/pkg/project/project_utils.go:41]: I'm assuming `filePath` gives us the full absolute path while `zipFile.Name()` was only giving us the filename?  I would just double check that this is honored downstream of the caller.

**Review by wbreza — APPROVED** [review]
Added a couple minor comments & questions but otherwise looks good.

**Comment by v-xuto** [comment]
> @v-xuto - Can you please run manual regression / smoke tests on all the azdev ToDo templates and ensure that everything passes?
> 
@jongio We have finished the manual regression tests. We filed four new issues: https://github.com/Azure/azure-dev/issues/4346 ,https://github.com/Azure/azure-dev/issues/4347, https://github.com/Azure/azure-dev/issues/4348, https://github.com/Azure/azure-dev/issues/4349, and there is an old issues https://github.com/Azure/azure-dev/issues/3520.

File: cli/azd/pkg/project/project_utils.go (hunk 5 above)
Comment (jongio) [reply cli/azd/pkg/project/project_utils.go:60]: > @ellismg had implemented `.dockerignore` recently

@ellismg - Are you thinking we migrate the pack_source.go implementation to use the new dotignore support in this PR?

Comment (weikanglim) [reply cli/azd/pkg/project/project_utils.go:60]: > .funcignore only honors files in the root and in discussion with them and the app service team we decided to not honor recursive .ignore files.

That is quite an interesting decision. From a dev perspective, one main reason why `*.ignore` files are used today is precisely because they can always be placed relative to a directory and provide a more local ignore scope. If that is not being done, a simpler package manifest at the root of the package directory would be a more ideal format IMO.

I took a closer look. It also turns out that `.funcignore` does not conform to the `.gitignore` spec entirely. @ellismg @jongio 

- The issue on [functools](https://github.com/Azure/azure-functions-core-tools/issues/2701)
- Based on the previous issue, I gather that the library they're using is [gitignore-parser](https://github.com/codemix/gitignore-parser/issues) with its set of issues.

If we're already engaged in conversations with both app service and function teams, can we consider:

1. Evaluating the impact of having two different `.funcignore` interpretations (is there user pain we would worry about?)
2. Based on those conversations, decide if we would like to rally all teams to adopt the conformant spec, and think about potentially support recursive directories. We could also just support recursive directories from `azd` side, it would be perhaps 20 lines of code.

Comment (jongio) [reply cli/azd/pkg/project/project_utils.go:60]: Clarifying that I meant to say..."service root", not the root of the repo.

This implementation honors .ignore files in the root of the service.  

I think we are on the same page.

Comment (jongio) [reply cli/azd/pkg/project/project_utils.go:60]: @weikanglim - Any objections to moving forward with the current implementation as it matches what .funcignore does today and what they would like for .webappignore.  

This means a service zipping only recognizes the .ignore file that is in the root of that service folder and not any parent .ignore files.

File: cli/azd/pkg/project/project_utils.go (hunk 6 above)
Comment (jongio) [reply cli/azd/pkg/project/project_utils.go:41]: I have confirmed that both filePath and zipFile.Name() return the same value:

image

File: cli/azd/pkg/project/project_utils.go (hunk 5 above)
Comment (weikanglim) [reply cli/azd/pkg/project/project_utils.go:60]: @jongio Just checking -- where are we in resolving the difference between of behavior in interpretation for `.funcignore` by func-tools vs. azd, as highlighted [previously](https://github.com/Azure/azure-dev/pull/4258/files#r1767506732)? I haven't seen it mentioned.

Aside from that:
1. I'm a little unhappy about `dotignore` being a package and the implementation of the functions being tied to working "within a directory". I suspect in the future this may end up needing slight rewriting to support recursive directories. I'm more than happy to work with you to push some changes here to clean up the implementation.
2. I'm looking at the PR again, and it seems like there are significant changes to `rzip.go` that I'm unsure how or if they're related. They should be separated out if it's a bug fix.
3. I'm also a little nervous about the e2e tests that we're doing inspection of the zip files -- `package` likely requires changes in the future that would require us to rethink these test cases. The heavy addition of many test data files to set up "scenario tests" of all the projects also seem rather heavyweight and will be slightly harder to maintain in the long term. It feels like simply testing the "exclusion conditions" would suffice for testing the behavior with/without `*.ignore` files. 
4. We should double check again what the default behavior of without `.funcignore` should be -- I believe it should minimally ignore `local.settings.json` to be inline with function tools, see docs [here](https://learn.microsoft.com/en-us/azure/azure-functions/functions-run-local?tabs=macos%2Cisolated-process%2Cnode-v4%2Cpython-v2%2Chttp-trigger%2Ccontainer-apps&pivots=programming-language-java#upload-local-settings-to-azure).

Overall, I really do love the direction here. But I would love if we can have a simple clean landing of a minimal change -- there are currently 37 file changes in this PR alone. I'm more than happy to work with you to land a minimal change if we can document all the necessary requirements in a new issue referencing #1039 so we're all on the same page, and also to serve as official documentation for the feature.

**Comment by lcephas** [comment]
Hi folks. I came here from https://github.com/Azure/azure-dev/issues/4215, because I ran into the issue myself. My question is regarding the defaults. What's the reason not to include node_modules by default and make .webappignore the custom if you want to use the SCM build instead? IMO, this would make it more inline with other languages supported by AZD and is the "expected" behavior.

**Comment by jongio** [comment]
> Hi folks. I came here from #4215, because I ran into the issue myself. My question is regarding the defaults. What's the reason not to include node_modules by default and make .webappignore the custom if you want to use the SCM build instead? IMO, this would make it more inline with other languages supported by AZD and is the "expected" behavior.

@lcephas - Right now the default is to not include them, so instead of introducing unexpected behavior of including them by default and breaking all azd deployments with appservice or functions, we decided the best route is to maintain the existing behavior by default and allow a developer to break out of that behavior by using a custom .*ignore file.

File: cli/azd/pkg/project/project_utils.go (hunk 5 above)
Comment (jongio) [reply cli/azd/pkg/project/project_utils.go:60]: **On dotignore and recursive directories:**

Thanks for sharing your thoughts on dotignore. The current non-recursive behavior is consistent with how .funcignore works today, and I believe it's important to avoid introducing more differences. I’d prefer to maintain this alignment to keep things predictable and straightforward for developers.

**On changes to rzip.go:**

The changes in rzip.go are required to handle symbolic links in node_modules or virtual environments. Since these components could now be included, special code is necessary to ensure correct behavior. Keeping these updates within this PR ensures everything works seamlessly together.

**On e2e tests and test data files:**

Our current e2e tests are designed to cover any issues arising from the zip changes, particularly around symbolic links. The test data may be heavier, but it provides comprehensive coverage that ensures we don’t miss any potential edge cases.

**On default behavior for .funcignore:**

Regarding the behavior of .funcignore, I’m in touch with the functions team to verify if there are additional files that should be ignored, with or without the existence of a .funcignore file. It's also important to note that any issues where .funcignore doesn’t behave like .gitignore should be resolved within the func implementation. I don't think we should introduce behavior here that diverges from the expected .gitignore syntax.


Thank you for the offer, but I’m confident the current implementation is in line with our goals, and I don't feel any additional simplification is needed at this time. I appreciate your willingness to assist and your thoughtful feedback throughout this process

Comment (weikanglim) [reply cli/azd/pkg/project/project_utils.go:60]: I've left a comment on [rzip.go](https://github.com/Azure/azure-dev/pull/4258/files#r1775899423). It isn't clear what motivates all the changes here. I'd still push towards creating a separate issue for `rzip.go` change so that we can understand and track the changes separately.

> I’m in touch with the functions team to verify if there are additional files that should be ignored

Would we consider transferring all the existing documentation on the PR into an issue so that we can have the requirements discussion? I'm also more than happy to help here if needed.

Overall, I do really want this to see these changes landed -- I'm just trying to make sure we're covering enough our bases from enough angles to make this a good user experience for our users.

**Review by weikanglim — CHANGES_REQUESTED** [review]
Overall, I do really love the direction here and want to see these changes landed.

The small things I'm waiting on: I would love to see if we can land `rzip.go` separately, and also have the great documentation we have on this PR pushed upwards into an issue. Just let me know if you'd like any help with that.

File: cli/azd/pkg/project/project_utils.go (hunk 5 above)
Comment (jongio) [reply cli/azd/pkg/project/project_utils.go:60]: Here are the list of default excludes for function service type:
https://github.com/Azure/azure-functions-core-tools/blob/37b826f143b99cd62308b88320e71aa806f6197f/src/Azure.Functions.Cli/Common/FileSystemHelpers.cs#L144

We may also want to confirm with app service team if there are default excludes like appsettings.development.json

File: cli/azd/pkg/rzip/rzip.go [hunk 7]
```diff
@@ -12,47 +12,77 @@ import (
...
 
-		header := &zip.FileHeader{
-			Name: strings.Replace(
-				strings.TrimPrefix(
-					strings.TrimPrefix(path, source),
-					string(filepath.Separator)), "\\", "/", -1),
-			Modified: fileInfo.ModTime(),
-			Method:   zip.Deflate,
+		// Skip symbolic links
+		if fileInfo.Mode()&os.ModeSymlink != 0 {
```
[the remark this answers is not in the document]
Comment (jongio) [reply cli/azd/pkg/rzip/rzip.go:33]: I'm actually thinking that maybe we want to keep venvs excluded by default.  I don't think we have a good use case for including venvs in a zip folder.

@pamelafox - Can you confirm this?  And also do you think that we should include __pycache__ by default if they have a .ignore file?  I'm thinking it will be large and might not be needed in any scenario.
